NKL/Oper.lean

```diff
@@ -0,0 +1,181 @@
+/-
+Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Paul Govereau
+-/
+
+import NKL.NKI
+
+namespace NKL
+
+-- Some experimental definitions of semantics for expresssions
+-- For now, I am abstracting over the types of tensors and
+-- tensor operators.
+
+-- Evaluating expressions results in values, which
+-- may contain tensors and tensor operators
+inductive Value (tensor op : Type) where
+  | const (c : Const)
+  | tensor (t : tensor) (ix : List (Value tensor op))
+  | slice (l u step : Value tensor op)
+  | oper (f : op)
+  | tuple (l : List (Value tensor op))
+  | list (l : List (Value tensor op))
+  deriving Repr, BEq
+
+-- The semantics is defined w.r.t. an environment that defines
+-- the types of tensors and tensor operators, and provides
+-- a lookup function for getting the values of variables.
+class Env (a : Type) where
+  Tensor : Type
+  Oper : Type
+  lookup? : a -> String -> Option (Value Tensor Oper)
+  apply : a -> Oper -> List (Value Tensor Oper) -> Option (Value Tensor Oper)
+
+-- A convenient abbreviation for Value types
+abbrev Val (a :Type) [Env a] := Value (Env.Tensor a) (Env.Oper a)
+
+-- TODO: should probably rename "nil" to something else
+def nil [Env E]: Val E := .const .nil
+
+def Env.lookup [Env a] (x : a) (s : String) : Val a :=
+  Option.getD (Env.lookup? x s) nil
+
+-- Operational semantics of Expressions
+-- gridcall not included: this likely needs to be a statement
+
+mutual
+inductive StepExpr [Env E] : (env : E) -> Expr -> Val E -> Prop
+ | bvar :
+     Env.lookup? env a = some v ->
+     StepExpr env (.bvar a) v
+ | var :
+     Env.lookup env a = v ->
+     StepExpr env (.var a b) v
+ | subscript :
+     StepExpr env e (.tensor t []) ->
+     StepList env le lv ->
+     StepExpr env (.subscript e le) (.tensor t lv)
+ | slice :
+     StepExpr env l l' ->
+     StepExpr env u u' ->
+     StepExpr env step step' ->
+     StepExpr env (.slice l u step) (.slice l' u' step')
+ | binop :
```

Comment:
Why do we have both `binop` and `call`?

Reply:
Another left over from python syntax. We could translate the `binop` to `call` and the `gridcall` to `forloop` and `call`... and `cond` can be eliminated, as can `var` and `check`. More arguments for a "core language".